src/coreclr/jit/importercalls.cpp

```diff
@@ -4973,60 +4973,74 @@ GenTree* Compiler::impIntrinsic(CORINFO_CLASS_HANDLE    clsHnd,
+                        {
+                            op2 = impCloneExpr(op2, &op2Clone, CHECK_SPILL_ALL,
+                                               nullptr DEBUGARG("Clone op2 for Math.Min/Max non-Number"));
+                        }
+                        op1 = impCloneExpr(op1, &op1Clone, CHECK_SPILL_ALL,
+                                           nullptr DEBUGARG("Clone op1 for Math.Min/Max"));
+                    }
 
-                if (!isNumber)
-                {
-                    op2 = impCloneExpr(op2, &op2Clone, CHECK_SPILL_ALL,
-                                       nullptr DEBUGARG("Clone op2 for Math.Min/Max non-Number"));
-                }
+                    static const CORINFO_CONST_LOOKUP nullEntry = {IAT_VALUE};
 
-                if (!isNumber)
-                {
-                    op1 = impCloneExpr(op1, &op1Clone, CHECK_SPILL_ALL,
-                                       nullptr DEBUGARG("Clone op1 for Math.Min/Max non-Number"));
-                }
+                    // Native RISC-V fmin/fmax instructions implement IEEE 754-2019 Minimum/MaximumNumber functions
+                    // which don't specify what kind of NaN is returned if both arguments are NaN. RISC-V returns quiet
+                    // canonical NaN in this case.
+                    ni = isMax ? NI_System_Math_MaxNative : NI_System_Math_MinNative;
+                    GenTree* minMax =
+                        new (this, GT_INTRINSIC) GenTreeIntrinsic(callType, op1, op2, ni, nullptr R2RARG(nullEntry));
 
-                static const CORINFO_CONST_LOOKUP nullEntry = {IAT_VALUE};
-                if (isMagnitude)
-                {
-                    op1 = new (this, GT_INTRINSIC)
-                        GenTreeIntrinsic(callType, op1, NI_System_Math_Abs, nullptr R2RARG(nullEntry));
-                    op2 = new (this, GT_INTRINSIC)
-                        GenTreeIntrinsic(callType, op2, NI_System_Math_Abs, nullptr R2RARG(nullEntry));
-                }
+                    if (!isNative)
+                    {
+                        // Make sure we return the NaN argument verbatim (if both are NaN, the first one), which is an
+                        // additional requirement for .NET Min/Max APIs on top of IEEE 754.
 
-                ni = isMax ? NI_System_Math_MaxNumber : NI_System_Math_MinNumber;
-                GenTree* minMax =
-                    new (this, GT_INTRINSIC) GenTreeIntrinsic(callType, op1, op2, ni, nullptr R2RARG(nullEntry));
+                        if (isNumber)
+                        {
+                            // Build expression:  isNumber(minMax) ? minMax : op1
+                            unsigned tmpResult =
+                                lvaGrabTemp(true DEBUGARG("Temp for min/max result in Math.Min/MaxNumber"));
+                            impStoreToTemp(tmpResult, minMax, CHECK_SPILL_NONE);
+                            minMax = gtNewLclvNode(tmpResult, callType);
```

Comment:
I believe that `fgMakeMultiUse` is "preferred" for importer. That is,
```cpp
GenTree* minMaxDup = fgMakeMultiUse(&minMax);
```

Reply:
In the importer `impCloneExpr` should be preferred. We generally want to avoid creating commas whenever possible, and during import we have the framework available to do that.

Reply:
@tomeksowi it'd be good to fix this to use `impCloneExpr` rather than grabbing the temp and storing to it by hand.

I think everything else looks good and we can get it merged after.